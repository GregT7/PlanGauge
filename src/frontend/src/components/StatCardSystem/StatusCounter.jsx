// src/components/StatCardSystem/StatusCounter.jsx
import CountElement from "./CountElement";

const DEFAULT_COUNT = { good: 0, moderate: 0, poor: 0 };

function StatusCounter({ statusCount = DEFAULT_COUNT }) {
  const valid =
    Number.isInteger(statusCount?.good) &&
    Number.isInteger(statusCount?.moderate) &&
    Number.isInteger(statusCount?.poor);

  const counts = valid ? statusCount : DEFAULT_COUNT;

  return (
    <div className="flex justify-center gap-5 items-center py-5" data-testid="StatusCounter">
      <CountElement styling={"bg-emerald-500"} count={counts.good} />
      <CountElement styling={"bg-amber-500"} count={counts.moderate} />
      <CountElement styling={"bg-rose-500"} count={counts.poor} />
    </div>
  );
}

export default StatusCounter;
