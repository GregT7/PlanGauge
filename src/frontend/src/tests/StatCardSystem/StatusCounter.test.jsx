import { expect, it, describe } from "vitest";
import { screen, render } from "@testing-library/react";
import { useState } from "react";
import StatusCounter from "@/components/StatCardSystem/StatusCounter";

describe("StatusCounter unit testing", () => {
    it("passes snapshot test", () => {
        const statusCount = {"good": 0, "moderate": 0, "poor": 0};
        const { container } = render(<StatusCounter statusCount={statusCount}/>);
        expect(container).toMatchSnapshot();
    })

    it("renders when no props are passed", () => {
        const { container } = render(<StatusCounter/>);
    })

    it("renders when an empty array is passed", () => {
        const { container } = render(<StatusCounter statusCount={{}}/>);
        expect(screen.getByTestId("StatusCounter")).toBeInTheDocument();
    })

    it("renders when a string is passed", () => {
        const { container } = render(<StatusCounter statusCount={"test 123"}/>);
        expect(screen.getByTestId("StatusCounter")).toBeInTheDocument();
    })

    it("renders when one of the keys is missing (good, moderate, poor)", () => {
        const badData = {'good': 0, 'moderate': 0}
        const { container } = render(<StatusCounter statusCount={badData}/>);
        expect(screen.getByTestId("StatusCounter")).toBeInTheDocument();
    })

    it("renders when one of the keys doesn't have a number as a value", () => {
        const badData = {'good': 0, 'moderate': 0, 'poor': 'random text'}
        const { container } = render(<StatusCounter statusCount={badData}/>);
        expect(screen.getByTestId("StatusCounter")).toBeInTheDocument();
    });

    
    it("render with bad props and check to see that x0 is present x3 times and correct styling is applied", () => {
        const badData = {'good': 0, 'moderate': 0}
        render(<StatusCounter statusCount={badData}/>);
        const counters = screen.getAllByTestId('circle-div');
        expect(counters.length).toEqual(3);
        
        const firstCircle = counters[0];
        const text = firstCircle.parentElement.querySelector('span').textContent;
        expect(text).toBe('x0');

        const styleRegex = /bg-amber-500|bg-emerald-500|bg-rose-500/
        expect(counters[0]).toHaveClass(styleRegex)
        expect(counters[1]).toHaveClass(styleRegex)
        expect(counters[2]).toHaveClass(styleRegex)
    })
});