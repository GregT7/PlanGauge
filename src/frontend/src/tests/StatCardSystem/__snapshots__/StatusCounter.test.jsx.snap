// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatusCounter unit testing > passes snapshot test 1`] = `
<div>
  <div
    class="flex justify-center gap-5 items-center py-5"
    data-testid="StatusCounter"
  >
    <div
      class="flex items-center gap-1"
    >
      <div
        class="w-12 h-12 rounded-full bg-emerald-500"
        data-testid="circle-div"
      />
      <span>
        <span>
          x0
        </span>
      </span>
    </div>
    <div
      class="flex items-center gap-1"
    >
      <div
        class="w-12 h-12 rounded-full bg-amber-500"
        data-testid="circle-div"
      />
      <span>
        <span>
          x0
        </span>
      </span>
    </div>
    <div
      class="flex items-center gap-1"
    >
      <div
        class="w-12 h-12 rounded-full bg-rose-500"
        data-testid="circle-div"
      />
      <span>
        <span>
          x0
        </span>
      </span>
    </div>
  </div>
</div>
`;
